_posts/2020-10-01-torch-network-from-scratch/torch_network_from_scratch.Rmd

```diff
@@ -0,0 +1,1076 @@
+
+```{r}
+t1$stride()
+```
+
+    [1] 2 1
+
+For `t2`, of shape `3x2`, the distance between column elements is the same, but the distance between rows is now 3:
+
+```{r}
+t2$stride()
+```
+
+    [1] 3 1
+
+While zero-copy operations are optimal, there are cases where they won't work.
+
+With `view()`, this can happen when a tensor was obtained via an operation -- other than `view()` itself -- that itself has already modified the *stride*. One example would be `transpose()`:
+
+```{r}
+t1 <- torch_tensor(rbind(c(1, 2), c(3, 4), c(5, 6)))
+t1
+t1$stride()
+
+t2 <- t1$t()
+t2
+t2$stride()
+```
+
+    torch_tensor 
+     1  2
+     3  4
+     5  6
+    [ CPUFloatType{3,2} ]
+
+    [1] 2 1
+
+    torch_tensor 
+     1  3  5
+     2  4  6
+    [ CPUFloatType{2,3} ]
+
+    [1] 1 2
+
+In `torch` lingo, tensors -- like `t2` -- that re-use existing storage (and just read it differently), are said not to be "contiguous"[^1]. One way to reshape them is to use `contiguous()` on them before. We'll see this in the next subsection.
+
+[^1]: Although the assumption may be tempting, "contiguous" does not correspond to what we'd call "contiguous in memory" in casual language.
+
+#### Reshape with copy
+
+In the following snippet, trying to reshape `t2` using `view()` fails, as it already carries information indicating that the underlying data should not be read in physical order.
+
+```{r}
+t1 <- torch_tensor(rbind(c(1, 2), c(3, 4), c(5, 6)))
+
+t2 <- t1$t()
+
+t2$view(6) # error!
+```
+
+    Error in (function (self, size)  : 
+      view size is not compatible with input tensor's size and stride (at least one dimension spans across two contiguous subspaces).
+      Use .reshape(...) instead. (view at ../aten/src/ATen/native/TensorShape.cpp:1364)
+
+However, if we first call `contiguous()` on it, a *new tensor* is created, which may then be (virtually) reshaped using `view()`.[^2]
+
+[^2]: For correctness' sake, `contiguous()` will only make a copy if the tensor it is called on is *not contiguous already.*
+
+```{r}
+t3 <- t2$contiguous()
+
+t3$view(6)
+```
+
+    torch_tensor 
+     1
+     3
+     5
+     2
+     4
+     6
+    [ CPUFloatType{6} ]
+
+Alternatively, we can use `reshape()`. `reshape()` defaults to `view()`-like behavior if possible; otherwise it will create a physical copy.
+
+```{r}
+t2$storage()$data_ptr()
+
+t4 <- t2$reshape(6)
+
+t4$storage()$data_ptr()
+```
+
+    [1] "0x5648d49b4f40"
+
+    [1] "0x5648d2752980"
+
+### Operations on tensors
+
+Unsurprisingly, `torch` provides a bunch of mathematical operations on tensors; we'll see some of them in the network code below, and you'll encounter lots more when you continue your `torch` journey. Here, we just quickly take a look at the overall tensor method semantics.
```

Comment:
```suggestion
Unsurprisingly, `torch` provides a bunch of mathematical operations on tensors; we'll see some of them in the network code below, and you'll encounter lots more when you continue your `torch` journey. Here, we quickly take a look at the overall tensor method semantics.
```

Reply:
I use the word "just" a lot, too—I try and take it out when I can!

Reply:
defensive style again, right?